server.js

```diff
@@ -15,6 +17,121 @@ app.get("/", function (request, response) {
 
 // TODO add your routes and helper functions here
 
-const listener = app.listen(process.env.PORT, function () {
+app.get("/bookings", function (request, response) {
+  response.send(bookings);
+});
+
+
+app.get("/bookings/search", function (request, response) {
+  const datequery = request.query.date
+  const term = request.query.term
+  const date = moment(datequery, 'YYYY-MM-DD', true)
+
+  if (moment(date).isValid()) {
+    let result = bookings.filter(booking => moment(datequery).isBetween(booking.checkInDate, booking.checkOutDate) || moment(date).isSame(booking.checkInDate) || moment(date).isSame(booking.checkOutDate))
+    result.length > 0 ? response.send(result) : response.send('Not found a booking on this date')
+  } else if (datequery && !moment(date).isValid()) {
+    response.status(400).json('Please enter a valid date, format should like "YYYY-MM-DD"')
+  }
+  if (term) {
+    let result = bookings.filter((booking) => `${booking.firstName} ${booking.surname} ${booking.email}`.toLowerCase().includes(term.toLowerCase()))
+    result.length > 0 ? response.send(result) : response.send('Not found a booking with this term')
+  } else {
+    response.status(400).json('Bad request')
+  }
+});
+
+
+app.get("/bookings/:id", function (request, response) {
+  const id = Number(request.params.id)
+  if (!isNaN(id)) {
+    const booking = bookings.find(booking => booking.id === id)
+    if (booking) {
```

Comment:
This is a bit over engineering
creating two variable and doing two loops is not recommended. 
Ids should be primary key (unique ) so the filter is not needed, 
you can check the index and then call it directly